pandas/core/reshape.py

```diff
@@ -1280,7 +1320,10 @@ def get_empty_Frame(data, sparse):
     else:
         dummy_mat = np.eye(number_of_cols, dtype=np.uint8).take(codes, axis=0)
 
-        if not dummy_na:
+        if dummy_na == 'keep':
+            dummy_mat = dummy_mat.astype(np.float16)
```

Comment:
Yes that's right on both accounts. It is pretty straight forward to add more careful type handling if we think it is worth it. To be honest, it is not clear what should be done with types. 

Should allow multiple different types for each column being dummied?

Should we return float16 if any column has a nan and uint8 otherwise?

There is not a clear right answer afaict.